chainer/testing/link.py

```diff
@@ -0,0 +1,549 @@
+import unittest
+
+import numpy
+
+import chainer
+from chainer import backend
+from chainer import initializers
+from chainer.testing import array as array_module
+from chainer.testing import test
+
+
+class LinkTestError(test.TestError):
+    """Raised when the target link is implemented incorrectly."""
+    pass
+
+
+class InitializerPair(object):
+
+    """Class to hold a pair of initializer-like objects.
+
+    The initializer-like objects can be accessed via the ``first`` and the
+    ``second`` attributes.
+
+    When implementing ``LinkTestCase``, instances of this class can be included
+    in lists listing all initializer-like objects that should be tested in the
+    initializers test.
+    In that case, the first element should correspond to the initializer-like
+    argument passed to the constructor of the link, and the second element
+    correspond to the actual initializer-like object used by the link.
+    In many cases, they are the same, except when passing ``None`` to the
+    constructor.
+    """
+
+    def __init__(self, first, second):
+        initializers._check_is_initializer_like(first)
+        initializers._check_is_initializer_like(second)
+
+        self.first = first
+        self.second = second
+
+
+class LinkTestCase(unittest.TestCase):
+
+    """A base class for link test cases.
+
+    Link test cases can inherit from this class to define a set of link tests.
+
+    .. rubric:: Required methods
+
+    Each concrete class must at least override the following methods.
+
+    ``generate_inputs(self)``
+        Returns a tuple of input arrays of type :class:`numpy.ndarray`.
+
+    ``create_link(self, initializers)``
+        Returns a link. The link is typically initialized with the given
+        initializer-likes ``initializers``. ``initializers`` is a tuple of
+        same length as the number of parameters and contains initializer-likes
+        returned by either ``generate_forward_backward_initializers`` or
+        ``generate_initializers`` depending on the test being run.
+
+    .. rubric:: Optional methods
+
+    Additionally, the concrete class can override the following methods. Some
+    must be overridden depending on the skip flags  ``skip_forward_test``,
+    ``skip_backward_test`` and ``skip_initializers_test``.
+
+    ``generate_forward_backward_initializers(self)``
```

Comment:
I feel `generate_forward_backward_initializers` is too long and not easily understood to test implementors.

Reply:
How about renaming it to `generate_initializers` and change the name of the existing `generate_initializers` to something else, preferably something that is less "stochastic" such as `create_initializers`? It's difficult to convey the difference without making the names longer (so, assuming users will read the docs).

Reply:
Discussed: `generate_params` and `get_initializers` respectively.

Reply:
Fixed as discussed.